File: src/me/srinjoy/Events.java
package me.srinjoy;

import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Events extends ListenerAdapter {
    static String channel;
    @Override
    public void onGuildMessageReceived(@NotNull GuildMessageReceivedEvent event) {
        channel = event.getChannel().getId();
        String message = event.getMessage().getContentRaw();
        if (message.contains("-java") && message.contains("```")){
            Object[] obj = format(message.replace("-java", "").trim());
            System.out.println("Class:" + obj[0]);
            System.out.println(obj[1]);
            Class<?> clazz = new Execute(obj[0].toString(), obj[1].toString()).execute();
            if (obj[2].equals(true)) {
                try {
                    clazz.getDeclaredMethod("main", String[].class).invoke(clazz.newInstance(), new Object[]{null});
                    Field stringBuffer = clazz.getDeclaredField("stringBuffer");
                    event.getChannel().sendMessage((StringBuilder)stringBuffer.get(clazz.newInstance())).queue();
                    stringBuffer.set(clazz.newInstance(), new StringBuilder());
                }
                catch (NoSuchMethodException e)
                    { event.getChannel().sendMessage("Constructor matching " + obj[0]+"() not found").queue(); e.printStackTrace(); }
                catch (IllegalAccessException e)
                    { event.getChannel().sendMessage("class "+obj[0]+" is not public").queue(); e.printStackTrace(); }
                catch (InvocationTargetException | InstantiationException | NoSuchFieldException e) {  e.printStackTrace();  }
            } else event.getChannel().sendMessage("main not found").queue();
        }
    }
    private Object[] format(String message){
        String Class  = "";boolean main = false;
        message = message.substring(message.indexOf("```") + 3, message.lastIndexOf("```"));
        message = (message.contains("java") && message.indexOf("java") == 0)? message.replace("java", "").trim(): message;
        message = message.substring(0, message.indexOf('{') +1)
                    + "\npublic static StringBuilder stringBuffer = new StringBuilder();"
                    + message.substring(message.indexOf('{') +1);
        StringBuilder stringBuilder = new StringBuilder();
        StringTokenizer stringTokenizer = new StringTokenizer(message, "\n");
        while (stringTokenizer.hasMoreTokens()){
            String code = stringTokenizer.nextToken();
            if (code.contains("public static StringBuilder stringBuffer = new StringBuilder();") && !message.contains("public "+Class+"()"))
                stringBuilder.append("\n").append("public ").append(Class).append("(){}").append("\n");
            if (code.contains("System.out.println")){
                String args = code.substring(code.indexOf("System.out.println") + "System.out.println".length());
                args = args.substring(1, args.lastIndexOf(')'));
                code = code.replace("System.out.println("+args+")", "stringBuffer.append("+args+").append(\"\\n\")");
            }
            if (code.contains("System.out.print")){
                String args = code.substring(code.indexOf("System.out.print") + "System.out.print".length());
                args = args.substring(1, args.lastIndexOf(')'));
                code = code.replace("System.out.print("+args+")", "stringBuffer.append("+args+")");
            }
            stringBuilder.append(code).append("\n");
            ArrayList<String> list = new ArrayList<>(Arrays.asList(code.split(" ")));
            if (list.contains("class")) {
                Class = list.get(list.indexOf("class") + 1);
                Class = (Class.contains("{"))? Class.replace('{', ' ').trim() : Class;
            }
            if (list.contains("void") && list.get(list.indexOf("void")+1).contains("main")) main = true;
        }
        return new Object[]{Class, stringBuilder, main};
    }
}

File: src/main/kotlin/me/srinjoy/Events.kt
package me.srinjoy

import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.joor.Reflect
import org.joor.ReflectException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.PrintStream

class Events : ListenerAdapter() {
    override fun onReady(event: ReadyEvent) {
        println("${event.jda.selfUser.name} is ready!!")
    }
    override fun onSlashCommand(event: SlashCommandEvent) {
        println(event.name)
        when (event.name) {
            "stop" -> {
                event.reply("shutting down ${event.jda.selfUser.asMention}").setEphemeral(true).queue()
                event.jda.shutdown()
            }
            "compile" -> {
                println("compiling...")
                var code = event.getOption("code")?.asString!!.run {
                    if (contains("class")) this else {
                        event.reply("unable to find main class").setEphemeral(true).queue()
                        return
                    }
                }
                val args = event.getOption("args")?.asString?.split(" ")
                event.getOption("stdin")?.asString?.let { System.setIn(ByteArrayInputStream(it.toByteArray())) }
                if (!code.contains("```")) {
                    event.reply("enclose the code in a code block").setEphemeral(true).queue()
                    return
                }
                println("executing $code")
                code = code.replace(if (code.contains("```java")) "```java" else "```", "")
                val name = code.run {
                    (indexOf("class") + 5).let {
                        substring(it, indexOf("{", it))
                    }.trim()
                }
                println("class name: $name")
                val out = System.out
                val baosOut = ByteArrayOutputStream()
                System.setOut(PrintStream(baosOut))

                try {
                    Reflect.compile(name, code).call("main", args?.toTypedArray())
                    val stdout = String(baosOut.toByteArray())
                    println("stdout = $stdout")
                    event.reply("```output: $stdout```").queue()
                } catch (reflect: ReflectException) {
                    reflect.message?.let {
                        event.reply(it).setEphemeral(true).queue()
                        println("stderr = $it")
                    }
                    reflect.printStackTrace()
                }

                System.setOut(out)
            }
        }
    }
}